File: bindings/ruby/test/test_distribution.rb
require 'minitest/autorun'
require_relative '../lib/cconfigspace'

class CConfigSpaceTestDistribution < Minitest::Test
  def test_from_handle_roulette
    areas = [ 1.0, 2.0, 1.0, 0.5 ]
    d = CCS::RouletteDistribution::new(areas: areas)
    d2 = CCS::Object::from_handle(d)
    assert_equal( d.class, d2.class )
  end

  def roulette_check(areas, d)
    sum = areas.reduce(:+)
    assert_equal( :CCS_OBJECT_TYPE_DISTRIBUTION, d.object_type )
    assert_equal( :CCS_DISTRIBUTION_TYPE_ROULETTE, d.type )
    assert_equal( :CCS_NUMERIC_TYPE_INT, d.data_type )
    assert_equal( 1, d.dimension )
    assert_equal( 4, d.num_areas )
    assert( d.areas.reduce(:+) > 0.999 )
    assert( d.areas.reduce(:+) < 1.001 )
    d.areas.each_with_index { |a, i|
      assert( a < areas[i] * 1.001 / sum && a > areas[i] * 0.999 / sum )
    }
    i = d.bounds
    assert_equal( :CCS_NUMERIC_TYPE_INT, i.type)
    assert_equal( 0, i.lower)
    assert_equal( 4, i.upper)
    assert( i.lower_included? )
    refute( i.upper_included? )
  end

  def test_create_roulette
    areas = [ 1.0, 2.0, 1.0, 0.5 ]
    d = CCS::RouletteDistribution::new(areas: areas)
    roulette_check(areas, d)
  end

  def _test_serialize_roulette(fmt)
    areas = [ 1.0, 2.0, 1.0, 0.5 ]
    dref = CCS::RouletteDistribution::new(areas: areas)
    buff = dref.serialize(format: fmt)
    d = CCS::deserialize(format: fmt, buffer: buff)
    roulette_check(areas, d)
  end

  def test_serialize_roulette_binary
    _test_serialize_roulette(:binary)
  end

  def test_serialize_roulette_json
    _test_serialize_roulette(:json)
  end

  def test_from_handle_normal
    d = CCS::NormalDistribution::Float.new
    d2 = CCS::Object::from_handle(d)
    assert_equal( d.class, d2.class )
  end

  def normal_check(d)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_NORMAL )
    assert( d.data_type == :CCS_NUMERIC_TYPE_FLOAT )
    assert( d.scale == :CCS_SCALE_TYPE_LINEAR )
    assert( d.dimension == 1 )
    assert( d.mu == 0.0 )
    assert( d.sigma == 1.0 )
    i = d.bounds
    assert_equal( :CCS_NUMERIC_TYPE_FLOAT, i.type )
    assert_equal( -Float::INFINITY, i.lower )
    assert_equal( Float::INFINITY, i.upper )
    refute( i.lower_included? )
    refute( i.upper_included? )
  end

  def test_create_normal
    d = CCS::NormalDistribution::Float.new
    normal_check(d)
  end

  def _test_serialize_normal(fmt)
    dref = CCS::NormalDistribution::Float.new
    buff = dref.serialize(format: fmt)
    d = CCS::deserialize(format: fmt, buffer: buff)
    normal_check(d)
  end

  def test_serialize_normal_binary
    _test_serialize_normal(:binary)
  end

  def test_serialize_normal_json
    _test_serialize_normal(:json)
  end

  def test_create_normal_int
    d = CCS::NormalDistribution::Int.new(mu: 2.0, sigma: 5.0)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_NORMAL )
    assert( d.data_type == :CCS_NUMERIC_TYPE_INT )
    assert( d.scale == :CCS_SCALE_TYPE_LINEAR )
    assert( d.dimension == 1 )
    assert( d.mu == 2.0 )
    assert( d.sigma == 5.0 )
  end

  def test_create_normal_float
    d = CCS::NormalDistribution::Float.new(mu: 2.0, sigma: 5.0)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_NORMAL )
    assert( d.data_type == :CCS_NUMERIC_TYPE_FLOAT )
    assert( d.scale == :CCS_SCALE_TYPE_LINEAR )
    assert( d.dimension == 1 )
    assert( d.mu == 2.0 )
    assert( d.sigma == 5.0 )
  end

  def test_sample_normal
    rng = CCS::Rng::new
    d = CCS::NormalDistribution::Float.new
    i = d.bounds
    v = d.sample(rng)
    assert( i.include?(v) )
    a = d.samples(rng, 100)
    assert_equal(100, a.size)
    a.each { |w|
      assert( i.include?(w) )
    }
  end

  def test_from_handle_uniform
    d = CCS::UniformDistribution::Float.new
    d2 = CCS::Object::from_handle(d)
    assert_equal( d.class, d2.class )
  end

  def uniform_check(d)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_UNIFORM )
    assert( d.data_type == :CCS_NUMERIC_TYPE_FLOAT )
    assert( d.scale == :CCS_SCALE_TYPE_LINEAR )
    assert( d.dimension == 1 )
    i = d.bounds
    assert_equal( :CCS_NUMERIC_TYPE_FLOAT, i.type )
    assert_equal( 0.0, i.lower )
    assert_equal( 1.0, i.upper )
    assert( i.lower_included? )
    refute( i.upper_included? )
  end

  def test_create_uniform
    d = CCS::UniformDistribution::Float.new
    uniform_check(d)
  end

  def _test_serialize_uniform(fmt)
    dref = CCS::UniformDistribution::Float.new
    buff = dref.serialize(format: fmt)
    d = CCS::deserialize(format: fmt, buffer: buff)
    uniform_check(d)
  end

  def test_serialize_uniform_binary
    _test_serialize_uniform(:binary)
  end

  def test_serialize_uniform_json
    _test_serialize_uniform(:json)
  end

  def test_create_uniform_float
    d = CCS::UniformDistribution::Float.new(lower: -1.0, upper: 1.0)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_UNIFORM )
    assert( d.data_type == :CCS_NUMERIC_TYPE_FLOAT )
    assert( d.scale == :CCS_SCALE_TYPE_LINEAR )
    assert( d.dimension == 1 )
    i = d.bounds
    assert_equal( :CCS_NUMERIC_TYPE_FLOAT, i.type)
    assert_equal( -1.0, i.lower)
    assert_equal(  1.0, i.upper)
    assert_equal( -1.0, d.lower)
    assert_equal(  1.0, d.upper)
    assert( i.lower_included? )
    refute( i.upper_included? )
  end

  def test_create_uniform_int
    d = CCS::UniformDistribution::Int.new(lower: 0, upper: 100)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_UNIFORM )
    assert( d.data_type == :CCS_NUMERIC_TYPE_INT )
    assert( d.scale == :CCS_SCALE_TYPE_LINEAR )
    assert( d.dimension == 1 )
    i = d.bounds
    assert_equal( :CCS_NUMERIC_TYPE_INT, i.type)
    assert_equal(   0, i.lower)
    assert_equal( 100, i.upper)
    assert_equal(   0, d.lower)
    assert_equal( 100, d.upper)
    assert( i.lower_included? )
    refute( i.upper_included? )
  end

  def test_oversampling_uniform_float
    d = CCS::UniformDistribution::Float.new(lower: -1.0, upper: 1.0)
    i = CCS::Interval::new(type: :CCS_NUMERIC_TYPE_FLOAT, lower: -0.2, upper: 0.2)
    assert(d.oversampling?(i))
    i = CCS::Interval::new(type: :CCS_NUMERIC_TYPE_FLOAT, lower: -0.2, upper: 2.0)
    assert(d.oversampling?(i))
    i = CCS::Interval::new(type: :CCS_NUMERIC_TYPE_FLOAT, lower: -2, upper: 2.0)
    refute(d.oversampling?(i))
  end

  def test_oversampling_uniform_int
    d = CCS::UniformDistribution::Int.new(lower: 0, upper: 100)
    i = CCS::Interval::new(type: :CCS_NUMERIC_TYPE_INT, lower: 5, upper: 50)
    assert(d.oversampling?(i))
    i = CCS::Interval::new(type: :CCS_NUMERIC_TYPE_INT, lower: 5, upper: 150)
    assert(d.oversampling?(i))
    i = CCS::Interval::new(type: :CCS_NUMERIC_TYPE_INT, lower: -5, upper: 150)
    refute(d.oversampling?(i))
  end

  def test_sample_uniform
    rng = CCS::Rng::new
    d = CCS::UniformDistribution::Float.new
    i = d.bounds
    v = d.sample(rng)
    assert( i.include?(v) )
    a = d.samples(rng, 100)
    assert_equal(100, a.size)
    a.each { |w|
      assert( i.include?(w) )
    }
  end

  def test_create_mixture
    distributions = [ CCS::UniformDistribution::Float.new(lower: -5.0, upper: 0.0), CCS::UniformDistribution::Float.new(lower: 0.0, upper: 2.0) ]
    d = CCS::MixtureDistribution::new(distributions: distributions)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_MIXTURE )
    assert( d.data_types == [:CCS_NUMERIC_TYPE_FLOAT] )
    assert( d.weights == [0.5, 0.5] )
    assert( d.distributions.collect(&:handle) == distributions.collect(&:handle) )
    d2 = CCS::Object::from_handle(d)
    assert_equal( d.class, d2.class )
  end

  def _test_serialize_mixture(fmt)
    distributions = [ CCS::UniformDistribution::Float.new(lower: -5.0, upper: 0.0), CCS::UniformDistribution::Float.new(lower: 0.0, upper: 2.0) ]
    dref = CCS::MixtureDistribution::new(distributions: distributions)
    buff = dref.serialize(format: fmt)
    d = CCS::deserialize(format: fmt, buffer: buff)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_MIXTURE )
    assert( d.data_types == [:CCS_NUMERIC_TYPE_FLOAT] )
    assert( d.weights == [0.5, 0.5] )
    distributions.each_with_index { |d2ref, i|
      d2 = d.distributions[i]
      assert_equal( d2ref.class, d2.class )
      assert_equal( d2ref.lower, d2.lower )
      assert_equal( d2ref.upper, d2.upper )
    }
  end

  def test_serialize_mixture_binary
    _test_serialize_mixture(:binary)
  end

  def test_serialize_mixture_json
    _test_serialize_mixture(:json)
  end

  def test_create_multivariate
    distributions = [ CCS::UniformDistribution::Float.new(lower: -5.0, upper: 0.0), CCS::UniformDistribution::Int.new(lower: 0, upper: 2) ]
    d = CCS::MultivariateDistribution::new(distributions: distributions)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_MULTIVARIATE )
    assert( d.data_types == [:CCS_NUMERIC_TYPE_FLOAT, :CCS_NUMERIC_TYPE_INT] )
    assert( d.distributions.collect(&:handle) == distributions.collect(&:handle) )
    d2 = CCS::Object::from_handle(d)
    assert_equal( d.class, d2.class )
  end

  def _test_serialize_multivariate(fmt)
    distributions = [ CCS::UniformDistribution::Float.new(lower: -5.0, upper: 0.0), CCS::UniformDistribution::Int.new(lower: 0, upper: 2) ]
    dref = CCS::MultivariateDistribution::new(distributions: distributions)
    buff = dref.serialize(format: fmt)
    d = CCS::deserialize(format: fmt, buffer: buff)
    assert( d.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d.type == :CCS_DISTRIBUTION_TYPE_MULTIVARIATE )
    assert( d.data_types == [:CCS_NUMERIC_TYPE_FLOAT, :CCS_NUMERIC_TYPE_INT] )
    distributions.each_with_index { |d2ref, i|
      d2 = d.distributions[i]
      assert_equal( d2ref.class, d2.class )
      assert_equal( d2ref.lower, d2.lower )
      assert_equal( d2ref.upper, d2.upper )
    }
  end

  def test_serialize_multivariate_binary
    _test_serialize_multivariate(:binary)
  end

  def test_serialize_multivariate_json
    _test_serialize_multivariate(:json)
  end

  def test_mixture_multidim
    distributions = [ CCS::UniformDistribution::Float.new(lower: -5.0, upper: 0.0), CCS::UniformDistribution::Int.new(lower: 0, upper: 2) ]
    d = CCS::MultivariateDistribution::new(distributions: distributions)
    d2 = CCS::MixtureDistribution::new(distributions: [d, d])
    assert( d2.object_type == :CCS_OBJECT_TYPE_DISTRIBUTION )
    assert( d2.type == :CCS_DISTRIBUTION_TYPE_MIXTURE )
    assert( d2.data_types == [:CCS_NUMERIC_TYPE_FLOAT, :CCS_NUMERIC_TYPE_INT] )
    assert( d2.weights == [0.5, 0.5] )
  end

  def test_parameters_sample
    h = CCS::NumericalParameter::Float.new(lower: -1.0, upper: 1.0)
    d = CCS::UniformDistribution::Float.new(lower: -1.0, upper: 1.0)
    vals = d.parameters_sample([h])
    assert_equal( 1, vals.length )
    assert_kind_of( Float, vals[0] )
  end

  def test_parameters_samples
    h = CCS::NumericalParameter::Float.new(lower: -1.0, upper: 1.0)
    d = CCS::UniformDistribution::Float.new(lower: -1.0, upper: 1.0)
    vals = d.parameters_samples([h], 10)
    assert_equal( 10, vals.length )
    vals.each { |v|
      assert_equal( 1, v.length )
      assert_kind_of( Float, v[0] )
    }
    assert_equal( [], d.parameters_samples([h], 0) )
  end
end

File: bindings/python/test/test_distribution.py
import unittest
import sys
sys.path.insert(1, '.')
sys.path.insert(1, '..')
import cconfigspace as ccs

class TestDistribution(unittest.TestCase):

  def test_from_handle_Roulette(self):
    d = ccs.RouletteDistribution(areas = [ 1.0, 2.0, 1.0, 0.5 ])
    d2 = ccs.Object.from_handle(d.handle)
    self.assertEqual( d.__class__, d2.__class__)
    self.assertEqual( d.handle.value, d2.handle.value)

  def test_create_roulette(self):
    areas = [ 1.0, 2.0, 1.0, 0.5 ]
    s = sum(areas)
    d = ccs.RouletteDistribution(areas = areas)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.ROULETTE, d.type )
    self.assertEqual( ccs.NumericType.INT, d.data_type )
    self.assertEqual( 1, d.dimension )
    a = d.areas
    self.assertTrue(  sum(a) > 0.999 )
    self.assertTrue(  sum(a) < 1.001 )
    for i in range(len(a)):
      self.assertTrue( a[i] < areas[i] * 1.001 / s )
      self.assertTrue( a[i] > areas[i] * 0.999 / s )
    i = d.bounds
    self.assertEqual( ccs.NumericType.INT, i.type)
    self.assertEqual( 0, i.lower )
    self.assertEqual( 4, i.upper )
    self.assertTrue( i.lower_included )
    self.assertFalse( i.upper_included )

  def test_serialize_roulette(self):
    areas = [ 1.0, 2.0, 1.0, 0.5 ]
    s = sum(areas)
    dref = ccs.RouletteDistribution(areas = areas)
    buff = dref.serialize()
    d = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.ROULETTE, d.type )
    self.assertEqual( ccs.NumericType.INT, d.data_type )
    self.assertEqual( 1, d.dimension )
    a = d.areas
    self.assertTrue(  sum(a) > 0.999 )
    self.assertTrue(  sum(a) < 1.001 )
    for i in range(len(a)):
      self.assertTrue( a[i] < areas[i] * 1.001 / s )
      self.assertTrue( a[i] > areas[i] * 0.999 / s )
    i = d.bounds
    self.assertEqual( ccs.NumericType.INT, i.type)
    self.assertEqual( 0, i.lower )
    self.assertEqual( 4, i.upper )
    self.assertTrue( i.lower_included )
    self.assertFalse( i.upper_included )


  def test_from_handle_normal(self):
    d = ccs.NormalDistribution.Float()
    d2 = ccs.Object.from_handle(d.handle)
    self.assertEqual( d.__class__, d2.__class__)
    self.assertEqual( d.handle.value, d2.handle.value)

  def test_create_normal(self):
    d = ccs.NormalDistribution.Float()
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.NORMAL, d.type )
    self.assertEqual( ccs.NumericType.FLOAT, d.data_type )
    self.assertEqual( ccs.ScaleType.LINEAR, d.scale )
    self.assertEqual( 1, d.dimension )
    self.assertEqual( 0.0, d.mu )
    self.assertEqual( 1.0, d.sigma )
    i = d.bounds
    self.assertEqual( ccs.NumericType.FLOAT, i.type )
    self.assertEqual( float('-inf'), i.lower )
    self.assertEqual( float('inf'), i.upper )
    self.assertFalse( i.lower_included )
    self.assertFalse( i.upper_included )

  def test_serialize_normal(self):
    dref = ccs.NormalDistribution.Float()
    buff = dref.serialize()
    d = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.NORMAL, d.type )
    self.assertEqual( ccs.NumericType.FLOAT, d.data_type )
    self.assertEqual( ccs.ScaleType.LINEAR, d.scale )
    self.assertEqual( 1, d.dimension )
    self.assertEqual( 0.0, d.mu )
    self.assertEqual( 1.0, d.sigma )
    i = d.bounds
    self.assertEqual( ccs.NumericType.FLOAT, i.type )
    self.assertEqual( float('-inf'), i.lower )
    self.assertEqual( float('inf'), i.upper )
    self.assertFalse( i.lower_included )
    self.assertFalse( i.upper_included )

  def test_create_normal_float(self):
    d = ccs.NormalDistribution.Float(mu = 2.0, sigma = 6.0)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.NORMAL, d.type )
    self.assertEqual( ccs.NumericType.FLOAT, d.data_type )
    self.assertEqual( ccs.ScaleType.LINEAR, d.scale )
    self.assertEqual( 1, d.dimension )
    self.assertEqual( 2.0, d.mu )
    self.assertEqual( 6.0, d.sigma )

  def test_create_normal_int(self):
    d = ccs.NormalDistribution.Int(mu = 2.0, sigma = 6.0)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.NORMAL, d.type )
    self.assertEqual( ccs.NumericType.INT, d.data_type )
    self.assertEqual( ccs.ScaleType.LINEAR, d.scale )
    self.assertEqual( 1, d.dimension )
    self.assertEqual( 2.0, d.mu )
    self.assertEqual( 6.0, d.sigma )

  def test_sample_normal(self):
    rng = ccs.Rng()
    d = ccs.NormalDistribution.Float()
    i = d.bounds
    v = d.sample(rng)
    self.assertTrue( i.contains(v) )
    a = d.samples(rng, 100)
    self.assertEqual( 100, len(a) )
    for v in a:
      self.assertTrue( i.contains(v) )

  def test_from_handle_uniform(self):
    d = ccs.UniformDistribution.Float(lower = 0.0, upper = 1.0)
    d2 = ccs.Object.from_handle(d.handle)
    self.assertEqual( d.__class__, d2.__class__)
    self.assertEqual( d.handle.value, d2.handle.value)

  def test_create_uniform(self):
    d = ccs.UniformDistribution.Float(lower = 0.0, upper = 1.0)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.UNIFORM, d.type )
    self.assertEqual( ccs.NumericType.FLOAT, d.data_type )
    self.assertEqual( ccs.ScaleType.LINEAR, d.scale )
    self.assertEqual( 1, d.dimension )
    i = d.bounds
    self.assertEqual( ccs.NumericType.FLOAT, i.type )
    self.assertEqual( 0.0, i.lower)
    self.assertEqual( 1.0, i.upper)
    self.assertEqual( 0.0, d.lower)
    self.assertEqual( 1.0, d.upper)
    self.assertTrue( i.lower_included )
    self.assertFalse( i.upper_included )

  def test_serialize_uniform(self):
    dref = ccs.UniformDistribution.Float(lower = 0.0, upper = 1.0)
    buff = dref.serialize()
    d = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.UNIFORM, d.type )
    self.assertEqual( ccs.NumericType.FLOAT, d.data_type )
    self.assertEqual( ccs.ScaleType.LINEAR, d.scale )
    self.assertEqual( 1, d.dimension )
    i = d.bounds
    self.assertEqual( ccs.NumericType.FLOAT, i.type )
    self.assertEqual( 0.0, i.lower)
    self.assertEqual( 1.0, i.upper)
    self.assertEqual( 0.0, d.lower)
    self.assertEqual( 1.0, d.upper)
    self.assertTrue( i.lower_included )
    self.assertFalse( i.upper_included )

  def test_create_uniform_float(self):
    d = ccs.UniformDistribution.Float(lower = -1.0, upper = 1.0)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.UNIFORM, d.type )
    self.assertEqual( ccs.NumericType.FLOAT, d.data_type )
    self.assertEqual( ccs.ScaleType.LINEAR, d.scale )
    self.assertEqual( 1, d.dimension )
    i = d.bounds
    self.assertEqual( ccs.NumericType.FLOAT, i.type )
    self.assertEqual( -1.0, i.lower)
    self.assertEqual(  1.0, i.upper)
    self.assertEqual( -1.0, d.lower)
    self.assertEqual(  1.0, d.upper)
    self.assertTrue( i.lower_included )
    self.assertFalse( i.upper_included )

  def test_create_uniform_int(self):
    d = ccs.UniformDistribution.Int(lower = 0, upper = 100)
    self.assertEqual( ccs.ObjectType.DISTRIBUTION, d.object_type )
    self.assertEqual( ccs.DistributionType.UNIFORM, d.type )
    self.assertEqual( ccs.NumericType.INT, d.data_type )
    self.assertEqual( ccs.ScaleType.LINEAR, d.scale )
    self.assertEqual( 1, d.dimension )
    i = d.bounds
    self.assertEqual( ccs.NumericType.INT, i.type )
    self.assertEqual(  0,  i.lower)
    self.assertEqual( 100, i.upper)
    self.assertEqual(   0, d.lower)
    self.assertEqual( 100, d.upper)
    self.assertTrue( i.lower_included )
    self.assertFalse( i.upper_included )

  def test_oversampling_uniform_float(self):
    d = ccs.UniformDistribution.Float(lower = -1.0, upper = 1.0)
    i = ccs.Interval(t = ccs.NumericType.FLOAT, lower = -0.2, upper = 0.2)
    self.assertTrue( d.is_oversampling(i) )
    i = ccs.Interval(t = ccs.NumericType.FLOAT, lower = -0.2, upper = 2.0)
    self.assertTrue( d.is_oversampling(i) )
    i = ccs.Interval(t = ccs.NumericType.FLOAT, lower = -2.0, upper = 2.0)
    self.assertFalse( d.is_oversampling(i) )

  def test_oversampling_uniform_int(self):
    d = ccs.UniformDistribution.Int(lower = 0, upper = 100)
    i = ccs.Interval(t = ccs.NumericType.INT, lower = 5, upper = 50)
    self.assertTrue( d.is_oversampling(i) )
    i = ccs.Interval(t = ccs.NumericType.INT, lower = 5, upper = 150)
    self.assertTrue( d.is_oversampling(i) )
    i = ccs.Interval(t = ccs.NumericType.INT, lower = -5, upper = 150)
    self.assertFalse( d.is_oversampling(i) )

  def test_sample_uniform(self):
    rng = ccs.Rng()
    d = ccs.UniformDistribution.Float(lower = 0.0, upper = 1.0)
    i = d.bounds
    v = d.sample(rng)
    self.assertTrue( i.contains(v) )
    a = d.samples(rng, 100)
    self.assertEqual( 100, len(a) )
    for v in a:
      self.assertTrue( i.contains(v) )

  def test_create_mixture(self):
    distributions = [ ccs.UniformDistribution.Float(lower = -5.0, upper = 0.0),
                      ccs.UniformDistribution.Float(lower =  0.0, upper = 2.0) ]
    d = ccs.MixtureDistribution(distributions = distributions)
    self.assertEqual( d.object_type, ccs.ObjectType.DISTRIBUTION )
    self.assertEqual( d.type, ccs.DistributionType.MIXTURE )
    self.assertEqual( d.data_types, [ccs.NumericType.FLOAT] )
    self.assertEqual( d.weights, [0.5, 0.5] )
    self.assertEqual( [x.handle.value for x in d.distributions], [x.handle.value for x in distributions] )
    d2 = ccs.Object.from_handle(d.handle)
    self.assertEqual( d.__class__, d2.__class__ )

  def test_serialize_mixture(self):
    distributions = [ ccs.UniformDistribution.Float(lower = -5.0, upper = 0.0),
                      ccs.UniformDistribution.Float(lower =  0.0, upper = 2.0) ]
    dref = ccs.MixtureDistribution(distributions = distributions)
    buff = dref.serialize()
    d = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( d.object_type, ccs.ObjectType.DISTRIBUTION )
    self.assertEqual( d.type, ccs.DistributionType.MIXTURE )
    self.assertEqual( d.data_types, [ccs.NumericType.FLOAT] )
    self.assertEqual( d.weights, [0.5, 0.5] )
    for i in [0,1]:
      d2ref = distributions[i]
      d2 = d.distributions[i]
      self.assertEqual( d2ref.__class__, d2.__class__ )
      self.assertEqual( d2ref.lower, d2.lower )
      self.assertEqual( d2ref.upper, d2.upper )

  def test_create_multivariate(self):
    distributions = [ ccs.UniformDistribution.Float(lower = -5.0, upper = 0.0),
                      ccs.UniformDistribution.Int(lower =  0, upper = 2) ]
    d = ccs.MultivariateDistribution(distributions = distributions)
    self.assertEqual( d.object_type, ccs.ObjectType.DISTRIBUTION )
    self.assertEqual( d.type, ccs.DistributionType.MULTIVARIATE )
    self.assertEqual( d.data_types, [ccs.NumericType.FLOAT, ccs.NumericType.INT] )
    self.assertEqual( [x.handle.value for x in d.distributions], [x.handle.value for x in distributions] )
    d2 = ccs.Object.from_handle(d.handle)
    self.assertEqual( d.__class__, d2.__class__ )

  def test_serialize_multivariate(self):
    distributions = [ ccs.UniformDistribution.Float(lower = -5.0, upper = 0.0),
                      ccs.UniformDistribution.Int(lower =  0, upper = 2) ]
    dref = ccs.MultivariateDistribution(distributions = distributions)
    buff = dref.serialize()
    d = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( d.object_type, ccs.ObjectType.DISTRIBUTION )
    self.assertEqual( d.type, ccs.DistributionType.MULTIVARIATE )
    self.assertEqual( d.data_types, [ccs.NumericType.FLOAT, ccs.NumericType.INT] )
    for i in [0,1]:
      d2ref = distributions[i]
      d2 = d.distributions[i]
      self.assertEqual( d2ref.__class__, d2.__class__ )
      self.assertEqual( d2ref.lower, d2.lower )
      self.assertEqual( d2ref.upper, d2.upper )

  def test_mixture_multidim(self):
    distributions = [ ccs.UniformDistribution.Float(lower = -5.0, upper = 0.0),
                      ccs.UniformDistribution.Int(lower =  0, upper = 2) ]
    d = ccs.MultivariateDistribution(distributions = distributions)
    d2 = ccs.MixtureDistribution(distributions = [d, d])
    self.assertEqual( d2.object_type, ccs.ObjectType.DISTRIBUTION )
    self.assertEqual( d2.type, ccs.DistributionType.MIXTURE )
    self.assertEqual( d2.data_types, [ccs.NumericType.FLOAT, ccs.NumericType.INT] )
    self.assertEqual( d2.weights, [0.5, 0.5] )


if __name__ == '__main__':
    unittest.main()
